### PR TITLE
[L0] Initial version of queue dispatcher 

### PR DESCRIPTION
**Pull request by igchor**

### CODE REVIEW

File: source/adapters/level_zero/queue.cpp [hunk 1]
```diff
@@ -593,11 +603,17 @@ UR_APIEXPORT ur_result_t UR_APICALL urQueueRetain(
   return UR_RESULT_SUCCESS;
 }
 
-UR_APIEXPORT ur_result_t UR_APICALL urQueueRelease(
-    ur_queue_handle_t UrQueue ///< [in] handle of the queue object to release
+UR_APIEXPORT ur_result_t UR_APICALL urQueueRetain(
```
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:606]: Aliasing API function names like this is probably going to be confusing.

File: source/adapters/level_zero/queue.cpp [hunk 2]
```diff
@@ -504,6 +504,16 @@ UR_APIEXPORT ur_result_t UR_APICALL urQueueCreate(
 
   UR_ASSERT(Context->isValidDevice(Device), UR_RESULT_ERROR_INVALID_DEVICE);
 
+  // optimized path for immediate, in-order command lists
+  if (v2::shouldUseQueueV2() && (Flags & UR_QUEUE_FLAG_SUBMISSION_IMMEDIATE) &&
+      Device->useImmediateCommandLists() &&
+      (Flags & UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE) == 0) {
+    *Queue = new ur_queue_handle_t_(
+        std::in_place_type<v2::ur_queue_immediate_in_order_t>, Context, Device,
```
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:512]: I have to say, this is an incredibly clunky way of defining a variant of an enum. Can we e.g., create a factory function that would be take a templated variant argument?

File: source/adapters/level_zero/queue.cpp [hunk 3]
```diff
@@ -687,17 +703,31 @@ UR_APIEXPORT ur_result_t UR_APICALL urQueueRelease(
...
   return UR_RESULT_SUCCESS;
 }
 
-UR_APIEXPORT ur_result_t UR_APICALL urQueueGetNativeHandle(
-    ur_queue_handle_t UrQueue, ///< [in] handle of the queue.
+UR_APIEXPORT ur_result_t UR_APICALL urQueueRelease(
+    ur_queue_handle_t UrQueue ///< [in] handle of the queue object to release
+) {
+  return UrQueue->Dispatcher.dispatch([&](auto Queue) {
+    if constexpr (std::is_same_v<decltype(Queue), ur_queue_handle_legacy_t>) {
```
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:714]: why just this function is implemented in the closure for the dispatcher?

File: source/adapters/level_zero/device.hpp [hunk 4]
```diff
@@ -201,4 +201,7 @@ struct ur_device_handle_t_ : _ur_object {
   ZeCache<struct ze_global_memsize> ZeGlobalMemSize;
   ZeCache<ZeStruct<ze_mutable_command_list_exp_properties_t>>
       ZeDeviceMutableCmdListsProperties;
+
+  // Global (per-adapter) Id of the device
+  size_t Id;
```
Comment (pbalcer) [inline source/adapters/level_zero/device.hpp:206]: this is unrelated.

File: source/adapters/level_zero/queue.cpp [hunk 5]
```diff
@@ -593,11 +603,17 @@ UR_APIEXPORT ur_result_t UR_APICALL urQueueRetain(
   return UR_RESULT_SUCCESS;
 }
 
-UR_APIEXPORT ur_result_t UR_APICALL urQueueRelease(
-    ur_queue_handle_t UrQueue ///< [in] handle of the queue object to release
+UR_APIEXPORT ur_result_t UR_APICALL urQueueRetain(
+    ur_queue_handle_t UrQueue ///< [in] handle of the queue object to get access
 ) {
-  auto Queue = Legacy(UrQueue);
+  return UrQueue->Dispatcher.dispatch(
```
Comment (pbalcer) [inline source/adapters/level_zero/queue.cpp:609]: I know we are trying to minimize changes and use static dispatch where possible, but looking at this right now, I think simple inheritance from a base abstract class might be much simpler.

File: source/adapters/level_zero/queue.cpp (hunk 1 above)
Comment (igchor) [reply source/adapters/level_zero/queue.cpp:606]: Alternatives are:
- add prefix to the function, e.g. this would become urQueueRetainImpl or something like that
- make those functions member functions of the proper handle type, i.e. we would call UrQueue->Retain()

File: source/adapters/level_zero/device.hpp (hunk 4 above)
Comment (igchor) [reply source/adapters/level_zero/device.hpp:206]: I've put this in a separate commit. Do you want me to move it to a separate PR?

File: source/adapters/level_zero/queue.cpp (hunk 5 above)
Comment (igchor) [reply source/adapters/level_zero/queue.cpp:609]: It might be true, if we want to implement each function for each queue variant then there are not many benefits of using a variant anyway (like template functions), Approach with inheritance would look like this: https://github.com/igchor/unified-runtime/tree/queue_v2_inheritance

File: source/adapters/level_zero/device.hpp (hunk 4 above)
Comment (pbalcer) [reply source/adapters/level_zero/device.hpp:206]: I don't think we should have this at all yet, since it's dead code.

File: source/adapters/level_zero/queue.cpp (hunk 5 above)
Comment (pbalcer) [reply source/adapters/level_zero/queue.cpp:609]: Hm, both approaches have merit, but I think my preference would be inheritance just because its simpler to grok.

Comment (igchor) [reply source/adapters/level_zero/queue.cpp]: Ok, I've just pushed a version with inheritance. The base class and API functions that operate on queue are now auto-generated. The script and templates for generation can be merged or not, but in either case, if we'd like to go back to the variant-approach for some reason, it should be pretty easy to do.

**Review by isaacault — APPROVED** [review]
`source/adapters/level_zero/image.cpp` LGTM.

**Review by pbalcer — APPROVED** [review]

**Review by nrspruit — APPROVED** [review]
